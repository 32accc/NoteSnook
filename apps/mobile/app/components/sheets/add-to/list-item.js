/*
This file is part of the Notesnook project (https://notesnook.com/)

Copyright (C) 2023 Streetwriters (Private) Limited

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

import React, { useState } from "react";
import { View } from "react-native";
import { useThemeStore } from "../../../stores/use-theme-store";
import { SIZE } from "../../../utils/size";
import { IconButton } from "../../ui/icon-button";
import { PressableButton } from "../../ui/pressable";
import Heading from "../../ui/typography/heading";
import Paragraph from "../../ui/typography/paragraph";
import { useSelectionContext } from "./context";
import { FilteredList } from "./filtered-list";
import { useItemSelectionStore } from "./store";

const SelectionIndicator = ({ item, hasNotes, selectItem, onPress }) => {
  const itemState = useItemSelectionStore((state) => state.itemState[item.id]);
  const multiSelect = useItemSelectionStore((state) => state.multiSelect);

  const isSelected = itemState === "selected";
  const isIntermediate = itemState === "intermediate";
  const isRemoved = !isSelected && hasNotes;
  const colors = useThemeStore((state) => state.colors);

  return (
    <IconButton
      size={22}
      customStyle={{
        marginRight: 5,
        width: 23,
        height: 23
      }}
      color={
        isRemoved
          ? colors.red
          : isIntermediate || isSelected
          ? colors.accent
          : colors.icon
      }
      onPress={() => {
        if (multiSelect) return selectItem();
        onPress?.(item);
      }}
      testID={
        isRemoved
          ? "close-circle-outline"
          : isSelected
          ? "check-circle-outline"
          : isIntermediate
          ? "minus-circle-outline"
          : "checkbox-blank-circle-outline"
      }
      name={
        isRemoved
          ? "close-circle-outline"
          : isSelected
          ? "check-circle-outline"
          : isIntermediate
          ? "minus-circle-outline"
          : "checkbox-blank-circle-outline"
      }
    />
  );
};

export const ListItem = ({
  item,
  index,
  icon,
  infoText,
  hasSubList,
  onPress,
  onScrollEnd,
  getListItems,
  style,
  type,
  sublistItemType,
  onAddItem,
  getSublistItemProps,
  hasHeaderSearch,
  onAddSublistItem,
  hasNotes,
  sheetRef
}) => {
  const { toggleSelection } = useSelectionContext();
  const multiSelect = useItemSelectionStore((state) => state.multiSelect);

  const colors = useThemeStore((state) => state.colors);
  const [expanded, setExpanded] = useState(false);

  function selectItem() {
    toggleSelection(item);
  }

  return (
    <View
      style={{
        overflow: "hidden",
        marginBottom: 10,
        ...style
      }}
    >
      <PressableButton
        onPress={() => {
          if (hasSubList) return setExpanded(!expanded);
          if (multiSelect) return selectItem();
          onPress?.(item);
        }}
        type={type}
        onLongPress={() => {
          useItemSelectionStore.getState().setMultiSelect(true);
          selectItem();
        }}
        customStyle={{
          height: style?.height || 50,
          width: "100%",
          alignItems: "flex-start"
        }}
      >
        <View
          style={{
            width: "100%",
            height: 50,
            justifyContent: "space-between",
            flexDirection: "row",
            alignItems: "center",
            paddingHorizontal: 12
          }}
        >
          <View
            style={{
              flexDirection: "row",
              alignItems: "center"
            }}
          >
            <SelectionIndicator
              hasNotes={hasNotes}
              onPress={onPress}
              item={item}
              selectItem={selectItem}
            />
            <View>
              {hasSubList && expanded ? (
                <Heading size={SIZE.md}>{item.title}</Heading>
              ) : (
                <Paragraph size={SIZE.sm}>{item.title}</Paragraph>
              )}

              {infoText ? (
                <Paragraph size={SIZE.xs} color={colors.icon}>
                  {infoText}
                </Paragraph>
              ) : null}
            </View>
          </View>

          <View
            style={{
              flexDirection: "row"
            }}
          >
            {onAddSublistItem ? (
              <IconButton
                name={"plus"}
                testID="add-item-icon"
                color={colors}
                size={SIZE.xl}
                onPress={() => {
                  onAddSublistItem(item);
                }}
              />
            ) : null}
            {icon ? (
              <IconButton
                name={icon(expanded).name}
                color={icon(expanded).color}
                size={icon(expanded).size || SIZE.xl}
                onPress={
                  hasSubList
                    ? () => setExpanded(!expanded)
                    : icon(expanded).onPress
                }
              />
            ) : null}
          </View>
        </View>
      </PressableButton>

      {expanded && hasSubList ? (
        <FilteredList
          nestedScrollEnabled
          data={getListItems(item)}
          keyboardShouldPersistTaps="always"
          keyboardDismissMode="none"
          onMomentumScrollEnd={onScrollEnd}
          style={{
            width: "95%",
            alignSelf: "flex-end",
            maxHeight: 250
          }}
          itemType={sublistItemType}
          hasHeaderSearch={hasHeaderSearch}
          renderItem={({ item, index }) => (
            <ListItem
              item={item}
              {...getSublistItemProps(item)}
              index={index}
              onScrollEnd={onScrollEnd}
            />
          )}
          onAddItem={onAddItem}
        />
      ) : null}
    </View>
  );
};
