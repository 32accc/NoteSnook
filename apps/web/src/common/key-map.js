/*
This file is part of the Notesnook project (https://notesnook.com/)

Copyright (C) 2023 Streetwriters (Private) Limited

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

import hotkeys from "hotkeys-js";
import { navigate } from "../navigation";
// import { store as themestore } from "../stores/theme-store";
import { GlobalKeyboard, selectSearchInput } from "../utils/keyboard";

const KEYMAP = [
  // {
  //   keys: ["command+n", "ctrl+n", "command+alt+n", "ctrl+alt+n"],
  //   description: "Create a new note",
  //   global: true,
  //   action: (e) => {
  //     e.preventDefault();
  //     hashNavigate("/notes/create", {
  //       addNonce: true,
  //       replace: true,
  //       notify: true,
  //     });
  //   },
  // },
  // {
  //   keys: [
  //     "command+shift+n",
  //     "ctrl+shift+n",
  //     "command+shift+alt+n",
  //     "ctrl+shift+alt+n",
  //   ],
  //   description: "Create a new notebook",
  //   global: true,
  //   action: (e) => {
  //     e.preventDefault();
  //     hashNavigate("/notebooks/create", {
  //       replace: true,
  //       notify: true,
  //     });
  //   },
  // },
  {
    keys: ["command+f", "ctrl+f"],
    description: "Search all notes",
    global: false,
    action: (e) => {
      if (e.target?.classList.contains("ProseMirror")) {
        selectSearchInput();
        return;
      }
      e.preventDefault();

      navigate("/search/notes");
    }
  }
  // {
  //   keys: ["alt+n"],
  //   description: "Go to Notes",
  //   global: false,
  //   action: (e) => {
  //     e.preventDefault();
  //     navigate("/notes");
  //   },
  // },
  // {
  //   keys: ["alt+b"],
  //   description: "Go to Notebooks",
  //   global: false,
  //   action: (e) => {
  //     e.preventDefault();
  //     navigate("/notebooks");
  //   },
  // },
  // {
  //   keys: ["alt+f"],
  //   description: "Go to Favorites",
  //   global: false,
  //   action: (e) => {
  //     e.preventDefault();
  //     navigate("/favorites");
  //   },
  // },
  // {
  //   keys: ["alt+t"],
  //   description: "Go to Tags",
  //   global: false,
  //   action: (e) => {
  //     e.preventDefault();
  //     navigate("/tags");
  //   },
  // },
  // {
  //   keys: ["alt+d"],
  //   description: "Go to Trash",
  //   global: false,
  //   action: (e) => {
  //     e.preventDefault();
  //     navigate("/trash");
  //   },
  // },
  // {
  //   keys: ["alt+s"],
  //   description: "Go to Settings",
  //   global: false,
  //   action: (e) => {
  //     e.preventDefault();
  //     navigate("/settings");
  //   },
  // },
  // {
  //   keys: ["command+d", "ctrl+d"],
  //   description: "Toggle dark/light mode",
  //   global: true,
  //   action: (e) => {
  //     e.preventDefault();
  //     themestore.get().toggleNightMode();
  //   },
  // },
];

export function registerKeyMap() {
  hotkeys.filter = function () {
    return true;
  };

  KEYMAP.forEach((key) => {
    hotkeys(
      key.keys.join(","),
      { element: key.global ? GlobalKeyboard : window },
      key.action
    );
  });
}
