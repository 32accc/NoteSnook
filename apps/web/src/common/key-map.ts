/*
This file is part of the Notesnook project (https://notesnook.com/)

Copyright (C) 2023 Streetwriters (Private) Limited

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

import hotkeys from "hotkeys-js";
import { GlobalKeyboard } from "../utils/keyboard";
import { useEditorStore } from "../stores/editor-store";
import { useStore as useSearchStore } from "../stores/search-store";
import { useEditorManager } from "../components/editor/manager";

const KEYMAP = [
  // {
  //   keys: ["command+n", "ctrl+n", "command+alt+n", "ctrl+alt+n"],
  //   description: "Create a new note",
  //   global: true,
  //   action: (e) => {
  //     e.preventDefault();
  //     hashNavigate("/notes/create", {
  //       addNonce: true,
  //       replace: true,
  //       notify: true,
  //     });
  //   },
  // },
  // {
  //   keys: [
  //     "command+shift+n",
  //     "ctrl+shift+n",
  //     "command+shift+alt+n",
  //     "ctrl+shift+alt+n",
  //   ],
  //   description: "Create a new notebook",
  //   global: true,
  //   action: (e) => {
  //     e.preventDefault();
  //     hashNavigate("/notebooks/create", {
  //       replace: true,
  //       notify: true,
  //     });
  //   },
  // },
  {
    keys: ["command+f", "ctrl+f"],
    description: "Search all notes",
    global: false,
    action: (e: KeyboardEvent) => {
      const isInEditor =
        e.target instanceof HTMLElement &&
        !!e.target?.closest(".editor-container");
      if (isInEditor) {
        const activeSession = useEditorStore.getState().getActiveSession();
        if (activeSession?.type === "readonly") {
          e.preventDefault();
          const editor = useEditorManager
            .getState()
            .getEditor(activeSession.id);
          editor?.editor?.startSearch();
        }
        return;
      }
      e.preventDefault();

      useSearchStore.setState({ isSearching: true, searchType: "notes" });
    }
  }
  // {
  //   keys: ["alt+n"],
  //   description: "Go to Notes",
  //   global: false,
  //   action: (e) => {
  //     e.preventDefault();
  //     navigate("/notes");
  //   },
  // },
  // {
  //   keys: ["alt+b"],
  //   description: "Go to Notebooks",
  //   global: false,
  //   action: (e) => {
  //     e.preventDefault();
  //     navigate("/notebooks");
  //   },
  // },
  // {
  //   keys: ["alt+f"],
  //   description: "Go to Favorites",
  //   global: false,
  //   action: (e) => {
  //     e.preventDefault();
  //     navigate("/favorites");
  //   },
  // },
  // {
  //   keys: ["alt+t"],
  //   description: "Go to Tags",
  //   global: false,
  //   action: (e) => {
  //     e.preventDefault();
  //     navigate("/tags");
  //   },
  // },
  // {
  //   keys: ["alt+d"],
  //   description: "Go to Trash",
  //   global: false,
  //   action: (e) => {
  //     e.preventDefault();
  //     navigate("/trash");
  //   },
  // },
  // {
  //   keys: ["alt+s"],
  //   description: "Go to Settings",
  //   global: false,
  //   action: (e) => {
  //     e.preventDefault();
  //     navigate("/settings");
  //   },
  // },
  // {
  //   keys: ["command+d", "ctrl+d"],
  //   description: "Toggle dark/light mode",
  //   global: true,
  //   action: (e) => {
  //     e.preventDefault();
  //     themestore.get().toggleNightMode();
  //   },
  // },
];

export function registerKeyMap() {
  hotkeys.filter = function () {
    return true;
  };

  KEYMAP.forEach((key) => {
    hotkeys(
      key.keys.join(","),
      {
        element: key.global
          ? (GlobalKeyboard as unknown as HTMLElement)
          : document.body
      },
      key.action
    );
  });
}
