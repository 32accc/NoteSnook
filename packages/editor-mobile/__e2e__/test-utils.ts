/*
This file is part of the Notesnook project (https://notesnook.com/)

Copyright (C) 2023 Streetwriters (Private) Limited

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

import {
  AndroidDevice,
  BrowserContext,
  TestInfo,
  _android as android,
  test
} from "@playwright/test";
import { Keyboard, SUPPORTED_KEYBOARDS, getKeyboards } from "./keyboard-utils";
import * as ip from "ip";
import { DeviceSize, getDeviceSize } from "./device-utils";

type AndroidTestArgs = {
  chrome: BrowserContext;
  device: AndroidDevice;
  baseURL: string;
  keyboards: Keyboard[];
  size: DeviceSize;
};

const SERVER_ADDRESS = `http://${ip.address("Wi-Fi", "ipv4")}:3000`;

export function androidTest(
  title: string,
  testFunction: (
    args: AndroidTestArgs,
    testInfo: TestInfo
  ) => Promise<void> | void
) {
  test(title, async ({ channel: _ }, testInfo) => {
    const [device] = await android.devices();
    if (!device) {
      console.error("Please connect an Android device or emulator.");
      return;
    }

    const keyboards = await getKeyboards(device);
    if (!keyboards.length) {
      console.error(
        "No supported keyboard found. Please install one of",
        SUPPORTED_KEYBOARDS
      );
      return;
    }

    const deviceSize = await getDeviceSize(device);

    await device.shell("am force-stop com.android.chrome");
    const chrome = await device.launchBrowser({});

    await testFunction(
      {
        chrome: chrome,
        device,
        keyboards,
        baseURL: SERVER_ADDRESS,
        size: deviceSize
      },
      testInfo
    );

    await device.shell("am force-stop com.android.chrome");
    await device.close();
  });
}
