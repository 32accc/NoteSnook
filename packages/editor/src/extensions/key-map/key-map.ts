/*
This file is part of the Notesnook project (https://notesnook.com/)

Copyright (C) 2023 Streetwriters (Private) Limited

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

import { Extension } from "@tiptap/core";
import { isInTable } from "@tiptap/pm/tables";
import { isListActive } from "../../utils/prosemirror";

export const KeyMap = Extension.create({
  name: "key-map",

  addKeyboardShortcuts() {
    return {
      Tab: ({ editor }) => {
        if (isListActive(editor) || isInTable(editor.state)) return false;
        return editor.commands.insertContent("\t");
      },
      "Shift-Tab": ({ editor }) => {
        if (isListActive(editor)) return false;
        return true;
      }
    };
  }
});
