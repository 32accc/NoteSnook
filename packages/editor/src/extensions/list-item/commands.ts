/*
This file is part of the Notesnook project (https://notesnook.com/)

Copyright (C) 2023 Streetwriters (Private) Limited

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

import { EditorState } from "prosemirror-state";
import { NodeType } from "prosemirror-model";
import { Editor } from "@tiptap/core";
import {
  findParentNodeOfType,
  hasParentNodeOfType
} from "../../utils/prosemirror";

// WORKAROUND: if we're at the start of a list item, we need to either
// backspace directly to an empty list item above, or outdent this node
export function onBackspacePressed(
  editor: Editor,
  name: string,
  type: NodeType
) {
  const { selection } = editor.state;
  const { empty, $from } = selection;

  if (
    !empty ||
    !isInside(name, type, editor.state) ||
    $from.parentOffset !== 0 ||
    !isFirstChildOfParent(editor.state) ||
    !editor.can().liftListItem(type)
  )
    return false;

  const isEmpty = isListItemEmpty(type, editor.state);

  if (isEmpty) {
    if (isFirstOfType(type, editor.state)) {
      const parentList = getListFromListItem(type, editor.state);
      if (!parentList) return false;
      if (parentList.childCount > 1) {
        return editor.commands.liftListItem(type);
      }
    }

    return editor.chain().joinBackward().joinBackward().run();
  } else if (isFirstOfType(type, editor.state)) {
    return editor.commands.liftListItem(type);
  } else {
    const block = findParentNodeOfType(type)(selection);
    if (block && block.start === $from.pos - 1) {
      // we have to run join backward twice because on the first join
      // the two list items are joined i.e., the editor just puts their
      // paragraphs next to each other. The next join merges the paragraphs
      // like it should be.
      return editor.chain().joinBackward().joinBackward().run();
    }
  }
  return false;
}

export function onArrowUpPressed(editor: Editor, name: string, type: NodeType) {
  const { selection } = editor.state;
  const { empty } = selection;

  if (!empty || !isFirstOfType(type, editor.state)) return false;
  const parentList = getListFromListItem(type, editor.state);
  if (editor.state.doc.firstChild === parentList)
    return editor.commands.insertContentAt(0, "<p></p>");
  return false;
}

function isInside(name: string, type: NodeType, state: EditorState) {
  const { $from } = state.selection;

  const node = type || state.schema.nodes[name];
  const { paragraph } = state.schema.nodes;

  return (
    hasParentNodeOfType(node)(state.selection) &&
    $from.parent.type === paragraph
  );
}

function isFirstChildOfParent(state: EditorState): boolean {
  const { $from } = state.selection;
  return $from.depth > 1
    ? $from.parentOffset === 0 || $from.index($from.depth - 1) === 0
    : true;
}

const isFirstOfType = (type: NodeType, state: EditorState) => {
  const block = findParentNodeOfType(type)(state.selection);
  if (!block) return false;
  const { pos } = block;
  const resolved = state.doc.resolve(pos);
  return !resolved.nodeBefore;
};

const getListFromListItem = (type: NodeType, state: EditorState) => {
  const block = findParentNodeOfType(type)(state.selection);
  if (!block) return undefined;
  const { pos } = block;
  const resolved = state.doc.resolve(pos);
  if (
    !resolved.parent.type.spec.group ||
    resolved.parent.type.spec.group?.indexOf("list") <= -1
  )
    return undefined;

  return resolved.parent;
};

function isListItemEmpty(type: NodeType, state: EditorState) {
  const block = findParentNodeOfType(type)(state.selection);
  if (!block) return false;
  const { node } = block;
  return !node.textContent.length;
}
