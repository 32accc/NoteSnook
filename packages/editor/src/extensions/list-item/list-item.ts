/*
This file is part of the Notesnook project (https://notesnook.com/)

Copyright (C) 2023 Streetwriters (Private) Limited

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

import { ListItem as TiptapListItem } from "@tiptap/extension-list-item";

export const ListItem = TiptapListItem.extend({
  addKeyboardShortcuts() {
    return {
      ...this.parent?.(),
      Tab: (props) => {
        const { editor } = props;
        const { state } = editor;
        const { selection } = state;
        const { $from } = selection;

        if ($from.parent.type.name === "codeblock") return false;

        return this.parent?.()?.Tab(props) || false;
      }
    };
  }
});
