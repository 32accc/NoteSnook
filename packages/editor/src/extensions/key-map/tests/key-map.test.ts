/*
This file is part of the Notesnook project (https://notesnook.com/)

Copyright (C) 2023 Streetwriters (Private) Limited

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

import { createEditor, h } from "@/test-utils";
import { test, expect } from "vitest";
import BulletList from "../../bullet-list";
import { ListItem } from "../../list-item";
import { joinUpWithLastListItem } from "../key-map";

test(`join up with last list item in a flat list`, async () => {
  const { editor } = createEditor({
    initialContent: h("div", [
      h("ul", [h("li", ["Hello"]), h("li", ["World"])]),
      h("p")
    ]).innerHTML,
    extensions: {
      bulletList: BulletList,
      listItem: ListItem
    }
  });

  editor.commands.setTextSelection(editor.state.doc.nodeSize - 2);
  joinUpWithLastListItem(editor);

  expect(editor.state.selection.$from.parent.textContent).toBe("World");
  expect(editor.state.doc.content.toJSON()).toMatchSnapshot();
});

test(`join up with last list item in a nested list`, async () => {
  const { editor } = createEditor({
    initialContent: h("div", [
      h("ul", [
        h("li", ["Hello"]),
        h("li", [
          "World",
          h("ul", [
            h("li", ["Hello"]),
            h("li", ["World", h("ul", [h("li", ["Hello"])])]),
            h("li", ["A very nested world"])
          ])
        ])
      ]),
      h("p")
    ]).innerHTML,
    extensions: {
      bulletList: BulletList,
      listItem: ListItem
    }
  });

  editor.commands.setTextSelection(editor.state.doc.nodeSize - 2);
  joinUpWithLastListItem(editor);

  expect(editor.state.selection.$from.parent.textContent).toBe(
    "A very nested world"
  );
  expect(editor.state.doc.content.toJSON()).toMatchSnapshot();
});

test(`join up with last list item in a nested list`, async () => {
  const { editor } = createEditor({
    initialContent: h("div", [
      h("ul", [
        h("li", ["Hello"]),
        h("li", [
          "World",
          h("ul", [
            h("li", ["Hello"]),
            h("li", ["World", h("ul", [h("li", ["Hello"])])]),
            h("li", ["A very nested world"])
          ])
        ])
      ]),
      h("p")
    ]).innerHTML,
    extensions: {
      bulletList: BulletList,
      listItem: ListItem
    }
  });

  editor.commands.setTextSelection(editor.state.doc.nodeSize - 2);
  joinUpWithLastListItem(editor);

  expect(editor.state.selection.$from.parent.textContent).toBe(
    "A very nested world"
  );
  expect(editor.state.doc.content.toJSON()).toMatchSnapshot();
});
