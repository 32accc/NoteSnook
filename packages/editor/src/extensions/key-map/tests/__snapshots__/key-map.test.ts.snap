// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`join up with last list item in a flat list 1`] = `
[
  {
    "attrs": {
      "listType": "",
    },
    "content": [
      {
        "content": [
          {
            "content": [
              {
                "text": "Hello",
                "type": "text",
              },
            ],
            "type": "paragraph",
          },
        ],
        "type": "listItem",
      },
      {
        "content": [
          {
            "content": [
              {
                "text": "World",
                "type": "text",
              },
            ],
            "type": "paragraph",
          },
        ],
        "type": "listItem",
      },
    ],
    "type": "bulletList",
  },
]
`;

exports[`join up with last list item in a nested list 1`] = `
[
  {
    "attrs": {
      "listType": "",
    },
    "content": [
      {
        "content": [
          {
            "content": [
              {
                "text": "Hello",
                "type": "text",
              },
            ],
            "type": "paragraph",
          },
        ],
        "type": "listItem",
      },
      {
        "content": [
          {
            "content": [
              {
                "text": "World",
                "type": "text",
              },
            ],
            "type": "paragraph",
          },
          {
            "attrs": {
              "listType": "",
            },
            "content": [
              {
                "content": [
                  {
                    "content": [
                      {
                        "text": "Hello",
                        "type": "text",
                      },
                    ],
                    "type": "paragraph",
                  },
                ],
                "type": "listItem",
              },
              {
                "content": [
                  {
                    "content": [
                      {
                        "text": "World",
                        "type": "text",
                      },
                    ],
                    "type": "paragraph",
                  },
                  {
                    "attrs": {
                      "listType": "",
                    },
                    "content": [
                      {
                        "content": [
                          {
                            "content": [
                              {
                                "text": "Hello",
                                "type": "text",
                              },
                            ],
                            "type": "paragraph",
                          },
                        ],
                        "type": "listItem",
                      },
                    ],
                    "type": "bulletList",
                  },
                ],
                "type": "listItem",
              },
              {
                "content": [
                  {
                    "content": [
                      {
                        "text": "A very nested world",
                        "type": "text",
                      },
                    ],
                    "type": "paragraph",
                  },
                ],
                "type": "listItem",
              },
            ],
            "type": "bulletList",
          },
        ],
        "type": "listItem",
      },
    ],
    "type": "bulletList",
  },
]
`;

exports[`join up with last list item in a nested list 2`] = `
[
  {
    "attrs": {
      "listType": "",
    },
    "content": [
      {
        "content": [
          {
            "content": [
              {
                "text": "Hello",
                "type": "text",
              },
            ],
            "type": "paragraph",
          },
        ],
        "type": "listItem",
      },
      {
        "content": [
          {
            "content": [
              {
                "text": "World",
                "type": "text",
              },
            ],
            "type": "paragraph",
          },
          {
            "attrs": {
              "listType": "",
            },
            "content": [
              {
                "content": [
                  {
                    "content": [
                      {
                        "text": "Hello",
                        "type": "text",
                      },
                    ],
                    "type": "paragraph",
                  },
                ],
                "type": "listItem",
              },
              {
                "content": [
                  {
                    "content": [
                      {
                        "text": "World",
                        "type": "text",
                      },
                    ],
                    "type": "paragraph",
                  },
                  {
                    "attrs": {
                      "listType": "",
                    },
                    "content": [
                      {
                        "content": [
                          {
                            "content": [
                              {
                                "text": "Hello",
                                "type": "text",
                              },
                            ],
                            "type": "paragraph",
                          },
                        ],
                        "type": "listItem",
                      },
                    ],
                    "type": "bulletList",
                  },
                ],
                "type": "listItem",
              },
              {
                "content": [
                  {
                    "content": [
                      {
                        "text": "A very nested world",
                        "type": "text",
                      },
                    ],
                    "type": "paragraph",
                  },
                ],
                "type": "listItem",
              },
            ],
            "type": "bulletList",
          },
        ],
        "type": "listItem",
      },
    ],
    "type": "bulletList",
  },
]
`;
